File: Java/Fundamentals/Language/OOP/InterfaceTest2.java
interface TaxPayer{

	int pin();

	double getAnnualIncome();

	//extension method - an added method whose implementation is shared by all classes
	//which already inherit from this interface
	default double getIncomeTax(int age){
		float tr = age < 60 ? 0.15f : 0.10f;
		double ti = getAnnualIncome() - 120000;
		return ti > 0 ? tr * ti : 0;
	}
}

class Consultant extends payroll.Employee implements TaxPayer{

	public Consultant(int h, float r){
		super(h, r);
	}

	public int pin(){
		return 1000000 + getId();
	}

	public double getAnnualIncome(){
		return 12 * getNetIncome() + 25000;
	}

}

class Dealer implements TaxPayer{
	
	private int id;
	private double sales;

	public Dealer(int i, double s){
		id = i;
		sales = s;
	}

	public int pin(){
		return id;
	}

	public double getAnnualIncome(){
		return 0.15 * sales;
	}
}

class InterfaceTest2{

	public static void main(String[] args){
		TaxPayer jill = new Consultant(190, 200);
		TaxPayer jack = new Dealer(234567, 4000000);
		System.out.printf("Jill's income tax: %.2f%n", jill.getIncomeTax(36));
		System.out.printf("Jack's income tax: %.2f%n", jack.getIncomeTax(63));
	}
}

File: MS.NET/Fundamentals/Language/OOP/interfacetest2.cs
using System;

interface ITaxPayer
{
	int PIN {get;}
	double GetAnnualIncome();
}

static class TaxPayers
{
	//extension method - a static method that can be invoked as an 
	//instance method of its first this qualified argument type
	public static double GetIncomeTax(this ITaxPayer that, int age)
	{
		double ex = that.GetAnnualIncome() - 120000;
		float tr = age < 60 ? 0.15f : 0.12f;
	
		return ex > 0 ? tr * ex : 0;
	}
}

class Consultant : Payroll.Employee, ITaxPayer
{
	internal Consultant(int h, float r) : base(h, r) {}

	public int PIN
	{
		get { return Id + 100001; }
	}

	public double GetAnnualIncome()
	{
		return 12 * GetNetIncome() + 25000;
	}	
}

class Dealer : ITaxPayer
{
	private int pin;
	private double sales;

	internal Dealer(int p, double s)
	{
		pin = p;
		sales = s;
	}

	public int PIN => pin;

	public double GetAnnualIncome()
	{
		return 0.2 * sales;
	}
}

static class Program
{
	public static void Main()
	{
		Consultant jill = new Consultant(190, 250);
		Dealer jack = new Dealer(202002, 5200000);
		
		Console.WriteLine("Jill's PIN is {0} and Tax is {1:0.00}", jill.PIN, jill.GetIncomeTax(36)); //TaxPayers.GetIncomeTax(jill, 36)
		Console.WriteLine("Jack's PIN is {0} and Tax is {1:0.00}", jack.PIN, jack.GetIncomeTax(63));
	}
}